packages/test/end-to-end-tests/src/test/upgradeManager.spec.ts

```diff
@@ -39,68 +39,58 @@ class TestDataObject extends DataObject {
 }
 
 describe("UpgradeManager (hot-swap)", () => {
-    const documentId = "upgradeManagerTest";
-    const documentLoadUrl = `fluid-test://localhost/${documentId}`;
     const codeDetails: IFluidCodeDetails = {
         package: "localLoaderTestPackage",
         config: {},
     };
-
-    let deltaConnectionServer: ILocalDeltaConnectionServer;
-    let urlResolver: LocalResolver;
-    let documentServiceFactory: LocalDocumentServiceFactory;
+    let driver: ITestDriver;
     let opProcessingController: OpProcessingController;
 
-    async function createContainer(factory: IFluidDataStoreFactory): Promise<IContainer> {
+    async function createContainer(documentId: string, factory: IFluidDataStoreFactory): Promise<IContainer> {
         const codeLoader: ICodeLoader = new LocalCodeLoader([[codeDetails, factory]]);
         const loader = new Loader({
-            urlResolver,
-            documentServiceFactory,
+            urlResolver: driver.createUrlResolver(),
+            documentServiceFactory: driver.createDocumentServiceFactory(),
             codeLoader,
             options: { hotSwapContext: true },
         });
 
         return createAndAttachContainer(
-            codeDetails, loader, urlResolver.createCreateNewRequest(documentId));
+            codeDetails, loader, driver.createCreateNewRequest(documentId));
     }
 
-    async function loadContainer(factory: IFluidDataStoreFactory): Promise<IContainer> {
+    async function loadContainer(documentId: string, factory: IFluidDataStoreFactory): Promise<IContainer> {
         const codeLoader: ICodeLoader = new LocalCodeLoader([[codeDetails, factory]]);
         const loader = new Loader({
-            urlResolver,
-            documentServiceFactory,
+            urlResolver: driver.createUrlResolver(),
+            documentServiceFactory: driver.createDocumentServiceFactory(),
             codeLoader,
             options: { hotSwapContext: true },
         });
 
-        return loader.resolve({ url: documentLoadUrl });
+        return loader.resolve({ url: driver.createContainerUrl(documentId) });
     }
 
     beforeEach(async () => {
-        deltaConnectionServer = LocalDeltaConnectionServer.create();
-        urlResolver = new LocalResolver();
-        documentServiceFactory = new LocalDocumentServiceFactory(deltaConnectionServer);
+        driver = getFluidTestDriver();
         opProcessingController = new OpProcessingController();
     });
 
-    afterEach(async () => {
-        await deltaConnectionServer.webSocketServer.close();
-    });
-
     it("prevents multiple approved proposals", async () => {
         const clients = 10;
 
         const addCounts = Array(clients).fill(0);
         const approveCounts = Array(clients).fill(0);
         const containers: IContainer[] = [];
+        const documentId = createDocumentId();
```

Comment:
> createDocumentId() [](start = 26, length = 19)

could also move this into the beforeEach

Reply:
i don't want to create a mutable variable at the higher scope if i don't need to